### ISSUE TITLE
[Bug / Request] Check deb Install Before Updating Config

### ISSUE THREAD

**Issue by zero77** [doonze/EmbyUpdate#3]
When checking the logs, i have found that even if the deb install failed for whatever reason. The config.ini would still be updated with the latest version, even if the current installed version is still the old one.

Is it possible to check that the deb installed successfully before updating the config.ini.
Thanks

**Comment by doonze** [comment]
Hmmmm... I'll check my logic. Any errors on install should throw an
exception and exit the script before updating the config file. That was the
original logic, and I didn't change that. Can you tell specifically where
it's failing so I can recreate?

I'll take a look tonight and see if there are any logic flaws. I'll also
look into more checks. I can find no documentation currently on a way to
get Emby itself to give me it's version number, but I'll revisit that as
well. Maybe ask Luke directly.

I'll let you know what I find.

**Comment by zero77** [comment]
On one occasion it failed the deb install when it was quite a few beta versions out of date.
Also, i am running it from a root cron and not manually, if that makes a difference.

I will post any error logs when i get them again.

**Comment by zero77** [comment]
@doonze 
This is one of the errors i got:
https://pastebin.com/raw/uCdsrbjh

**Comment by doonze** [comment]
Interesting. What distro are you running? It seems dpkg can't find all it's
parts on your system. I'll have to research what is causing that. I have
some ideas on how to flag that as failed, so at least it won't keep saying
install was successful. I'll get that fixed later this week. I'm more
interested in your error. dpkg is getting triggered, so it can find that on
the path. But can't find other internal parts it needs. That's just weird.

Does it just happen when running from Cron? It maybe a issue with the Cron
your using. Do you get the same errors using sudo or switching to root and
running manually??

**Comment by zero77** [comment]
Yes, it's a bit of a weird error, i am using ubuntu 18.04 and it only seems to happen when its run as a cron and not manually.
> 0 */2 * * * python /home/user/EmbyUpdate/embyupdate.py >> /home/user/EmbyUpdate/embyupdate2.log 2>&1

with out root

I think it may be unrelated but i am also getting this:
https://pastebin.com/raw/zxjLvpQa

**Comment by doonze** [comment]
I doubt anything other than root has access to run the commands needed. It will also depend on which crontab your using. If you don't want to run it as root, you'll have to figure out a way to get all the accesses needed for the user you are running as, and also make sure the crontab your using has all the correct paths. 

As a test, I suggest commenting out your current crontab entry, and instead run a `sudo nano /etc/crontab` (or vim or vi or whatever you use) and putting something like this line in:

`45 12   * * *   root    /usr/bin/python3 /home/justin/scripts/embyupdate/embyupdate.py >> /home/justin/scripts/embyupdate/embyupdate.log 2>&1`

Of course changeing as needed. Also make sure the PATH at the top of it has at least all these:

`PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin`

I'm pretty sure what your seeing is a combination of the user it's running under and path issues on the crontab your using. I'm old school and almost always use /etc/crontab over the others. I've just found it takes the least amount of work to get working correctly. 

All that being said, you could google what permissions/changes need to be made to get the user you want to be able to run the normally root only system commands needed, and you could modify the paths of the crontab your using to add the paths above. But "out of the box" this script is designed to run as root (or manually as sudo) in the main system-wide crontab. I'll help you all I can to run it how you want, but all my testing was done with the above setup.

In other news, I'm looking at the code tonight for picking up on your errors and correctly realizing there has been a failure, then exiting without updating the config file. I'll let you know when I have new code out to fix it.

**Comment by doonze** [comment]
Version 2.2 Beta released. This has a fix for the issue you encountered where the script updated the config even when it had failed to update due to errors. I had assumed if a sub process got an exception it would pass that exception back to the calling script. This is not the case in python. It now collects the sub process return code, and if it's anything other then 0 (meaning there were errors) it then lets you know and exits the script before updating the config. 

So that problem I think is solved. Now if we can just sort your path issues on your system we'll be golden!

Next time you manually run the script, or cron runs it, it SHOULD self update the script. However the changes won't reflect until the NEXT run after the update.  I have plans to fix the need for a second run before changes are reflected, but it's a bit of an overhaul in code logic. I want to get it stable before changing that much code.

**Comment by zero77** [comment]
@doonze 
Thanks for taking the time to look at this, i will do some research into this also, i will try out some different crons under root as you suggested.
Thanks for the amendments i think this will make debuging this problum a lot easier now.

The cron is currently run as root but the scripts and sub directories are stored within another user's home directory. Is it worth moving them to /root instead.

**Comment by doonze** [comment]
Shouldn't matter. Root has access to all. I'd put it in that Cron I pasted
above. Make sure you put root in the user column. And verify the path at
the top of it matches what I posted. From your error this is 100% a path
issue. I've found often the system-wide Cron has the best path setup.

I suppose I could turn this into a systemd task, and use an internal
scheduler, or the systemd scheduler. But that gets a lot more complicated.
Maybe down the road.
